### PR TITLE
dumpxml: --update-cpu with query qemu/kvm for feature support query

### PR DESCRIPTION
**Pull request by czhang0**
Since libvirt 3.9, libvirt query qemu/kvm to get one feature support
or not by qemu, all supported features are given by host model + features
list in domcapabilities.xml. host model is one set of feature lists
defined in /usr/lib/libvirt/cpu_map.xml

While on older version, libvirt directly query host /proc/cpuinfo to get
whether host cpu support one feature or not. If one feature is found, 
means it is supported, which is unbelievable when this feature can't
be simulated by qemu.

Change:
Query domcapabilites and cpu_map.xml to get one feature support
or not.

Signed-off-by: cuzhang <cuzhang@redhat.com>

### CODE REVIEW

**Comment by czhang0** [comment]
run test PASS
avocado run  --vt-type libvirt type_specific.local.virsh.dumpxml.normal_test.non_acl.vm_shutoff.with_cpu
JOB ID     : c27441c329c3301e3d6a88168271d65b6bbff835
JOB LOG    : /root/avocado/job-results/job-2018-01-22T18.25-c27441c/job.log
 (1/6) type_specific.local.virsh.dumpxml.normal_test.non_acl.vm_shutoff.with_cpu.minimum_match.domname: PASS (2.36 s)
 (2/6) type_specific.local.virsh.dumpxml.normal_test.non_acl.vm_shutoff.with_cpu.minimum_match.domuuid: PASS (2.33 s)
 (3/6) type_specific.local.virsh.dumpxml.normal_test.non_acl.vm_shutoff.with_cpu.exact_match.domname: PASS (2.33 s)
 (4/6) type_specific.local.virsh.dumpxml.normal_test.non_acl.vm_shutoff.with_cpu.exact_match.domuuid: PASS (2.30 s)
 (5/6) type_specific.local.virsh.dumpxml.normal_test.non_acl.vm_shutoff.with_cpu.strict_match.domname: PASS (2.32 s)
 (6/6) type_specific.local.virsh.dumpxml.normal_test.non_acl.vm_shutoff.with_cpu.strict_match.domuuid: PASS (2.29 s)
RESULTS    : PASS 6 | ERROR 0 | FAIL 0 | SKIP 0 | WARN 0 | INTERRUPT 0 | CANCEL 0
JOB TIME   : 17.12 s

File: libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py [hunk 1]
```diff
@@ -47,11 +49,27 @@ def custom_cpu(vm_name, cpu_match):
...
+        :return: list of feature string
+        """
+        features = []
+        dom_capa = domcapability_xml.DomCapabilityXML()
+        modelname = dom_capa.get_modelname()
+        for item in dom_capa.get_additional_feature_list('host-model'):
+            for key, value in item.items():
+                if value == 'require':
+                    features.append(key)
+        logging.debug("all additional required feature: %s" % type(features))
```
[the remark this answers is not in the document]
Comment (czhang0) [reply libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py]: It's merely to help to debug.

File: libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py [hunk 2]
```diff
@@ -107,10 +130,14 @@ def check_cpu(xml, cpu_match):
             # Count actual require features
             if f_policy == "require":
                 require_count += 1
+
         # Check
```
[the remark this answers is not in the document]
Comment (czhang0) [reply libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py]: Adopt.

File: libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (balamuruhans) [reply libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py]: okay, logging features might help but I believe logging type will not help as it is always List right ?

File: libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py [hunk 3]
```diff
@@ -47,11 +49,27 @@ def custom_cpu(vm_name, cpu_match):
         vmcpu_xml.add_feature('vmx', 'forbid')
         # Unsupport feature 'ia64'
         vmcpu_xml.add_feature('ia64', 'optional')
-        vmcpu_xml.add_feature('vme', 'require')
+        vmcpu_xml.add_feature('vme', 'optional')
         vmxml['cpu'] = vmcpu_xml
         logging.debug('Custom VM CPU: %s', vmcpu_xml.xmltreefile)
         vmxml.sync()
 
+    def get_features():
```
Comment (chunfuwen) [inline libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py]: rename get_features to get_cpu_features to make it more clear

File: libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py [hunk 4]
```diff
@@ -47,11 +49,27 @@ def custom_cpu(vm_name, cpu_match):
...
+        """
+        features = []
+        dom_capa = domcapability_xml.DomCapabilityXML()
+        modelname = dom_capa.get_hostmodel_name()
+        for item in dom_capa.get_additional_feature_list('host-model'):
+            for key, value in item.items():
+                if value == 'require':
+                    features.append(key)
+        logging.debug("all additional required features: %s" % type(features))
+        return features + utils_misc.get_model_features(modelname)
```
Comment (chunfuwen) [inline libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py]: "+" is not encourage to join two strings

File: libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py [hunk 5]
```diff
@@ -84,8 +102,13 @@ def check_cpu(xml, cpu_match):
             expect_policy = "disable"
             if f_name in ["xtpr", "vme", "ia64"]:
                 # Check if feature is support on the host
-                if host_capa.check_feature_name(f_name):
-                    expect_policy = "require"
+                # Since libvirt3.9, libvirt query qemu/kvm to get one feature support or not
+                if libvirt_version.version_compare(3, 9, 0):
+                    if f_name in get_features():
+                        expect_policy = "require"
+                else:
```
Comment (chunfuwen) [inline libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py]: I am a little concerning about else here. suppose version>3.9.0 and f_name not in get_features(), that means that there is no change to further check whether host_capa.check_feature_name(f_name):
+                        expect_policy = "require"

File: libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py (hunk 3 above)
Comment (czhang0) [reply libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py]: Adopt. Thanks!

File: libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py (hunk 4 above)
Comment (czhang0) [reply libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py]: '+' can work, but I'll modify to use dict.extend() to joint two lists.

File: libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py [hunk 6]
```diff
@@ -47,11 +49,28 @@ def custom_cpu(vm_name, cpu_match):
...
+        """
+        features = []
+        dom_capa = domcapability_xml.DomCapabilityXML()
+        modelname = dom_capa.get_hostmodel_name()
+        for item in dom_capa.get_additional_feature_list('host-model'):
+            for key, value in item.items():
+                if value == 'require':
+                    features.append(key)
+        logging.debug("all additional required features: %s" % type(features))
+        features.extend(utils_misc.get_model_features(modelname))
```
Comment (balamuruhans) [inline libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py]: can there be features in cpu_map.xml exist even in cpu/mode ? if so it is better to consider using `set()`

Comment (czhang0) [reply libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py]: For more robust, I adopt your advice like below:
    return list(set(features) | set(utils_misc.get_model_features(modelname)))

File: libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py [hunk 7]
```diff
@@ -47,11 +49,28 @@ def custom_cpu(vm_name, cpu_match):
...
+        :return: list of feature string
+        """
+        features = []
+        dom_capa = domcapability_xml.DomCapabilityXML()
+        modelname = dom_capa.get_hostmodel_name()
+        for item in dom_capa.get_additional_feature_list('host-model'):
+            for key, value in item.items():
+                if value == 'require':
+                    features.append(key)
+        logging.debug("all additional required features: %s" % type(features))
```
[the remark this answers is not in the document]
Comment (czhang0) [reply libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py]: I'll remove it, thanks.

File: libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py (hunk 6 above)
Comment (balamuruhans) [reply libvirt/tests/src/virsh_cmd/domain/virsh_dumpxml.py]: Looks good. :+1:

**Pull request merged**